scylla-cdc/src/stream_generations.rs

```diff
@@ -0,0 +1,407 @@
+    fn get_generation_by_timeuuid_query(&self) -> String {
+        format!(
+            r#"
+    SELECT time
+    FROM {}
+    WHERE key = 'timestamps'
+    AND time <= dateOf( ? )
+    LIMIT 1;
+    "#,
+            self.generations_table_name
+        )
+    }
+
+    // Internal function configurable for testing purposes.
+    pub async fn fetch_generation_by_timeuuid(
+        &self,
+        time: &Uuid,
+    ) -> Result<Option<Generation>, Box<dyn Error>> {
+        let result = self
+            .session
+            .query(self.get_generation_by_timeuuid_query(), (time,))
+            .await?
+            .rows;
+
+        GenerationFetcher::return_single_row(result)
+    }
+
+    // Function instead of constant for testing purposes.
+    fn get_next_generation_query(&self) -> String {
+        format!(
+            r#"
+    SELECT time
+    FROM {}
+    WHERE key = 'timestamps'
+    AND time > ?
+    ORDER BY time DESC
+    LIMIT 1;
+    "#,
+            self.generations_table_name
+        )
+    }
+
+    // Internal function configurable for testing purposes.
+    pub async fn fetch_next_generation(
+        &self,
+        generation: &Generation,
+    ) -> Result<Option<Generation>, Box<dyn Error>> {
+        let result = self
+            .session
+            .query(self.get_next_generation_query(), (Timestamp(generation.0),))
+            .await?
+            .rows;
+
+        GenerationFetcher::return_single_row(result)
+    }
+
+    // Function instead of constant for testing purposes.
+    fn get_stream_ids_by_time_query(&self) -> String {
+        format!(
+            r#"
+    SELECT streams
+    FROM {}
+    WHERE time = ?;
+    "#,
+            self.streams_table_name
+        )
+    }
+
+    // Internal function configurable for testing purposes.
+    pub async fn fetch_stream_ids(
+        &self,
+        timestamp: &Generation,
+    ) -> Result<Vec<StreamID>, Box<dyn Error>> {
+        let mut result_vec = vec![];
+
+        let mut rows = self
+            .session
+            .query_iter(
+                self.get_stream_ids_by_time_query(),
+                (Timestamp(timestamp.0),),
+            )
+            .await?
+            .into_typed::<(Vec<Vec<u8>>,)>();
+
+        while let Some(next_row) = rows.next().await {
+            let (ids,) = next_row?;
+            for id in ids {
+                result_vec.push(StreamID(id));
+            }
+        }
+        Ok(result_vec)
+    }
+
+    // Return single row containing generation.
+    pub fn return_single_row(row: Option<Vec<Row>>) -> Result<Option<Generation>, Box<dyn Error>> {
+        if let Some(row) = row {
+            return match row.into_typed::<(chrono::Duration,)>().next() {
+                None => Ok(None),
+                Some(row) => {
+                    let row = row?.0;
+                    Ok(Some(Generation(row)))
+                }
+            };
+        }
+
+        Ok(None)
+    }
+}
```

Comment:
Variable shadowing is cool but isn't using one name for four different types a bit too much? At least it's a bit confusing to me, especially because initially what you're refferring to as a singular `row` is actually a `Vec<Row>` (and yeah, I know its length is always 1 in your use cases but it still gives the opposite impression). What do you think @macher259 ? Maybe I am wrong and it's exactly what you're supposed to do in Rust? You probably know better.

I like it this way:
```rust
pub fn return_single_row(rows: Option<Vec<Row>>) -> Result<Option<Generation>, Box<dyn Error>> {
        if let Some(rows) = rows {
            return match rows.into_typed::<(chrono::Duration, )>().next() {
                None => Ok(None),
                Some(row) => {
                    let time = row?.0;
                    Ok(Some(Generation(time)))
                }
            };
        }

        Ok(None)
    }
```

Reply:
As a side note: you can shorten this to
```rust
pub fn return_single_row(row: Option<Vec<Row>>) -> Result<Option<Generation>, Box<dyn Error>> {
    if let Some(row) = row {
        if let Some(row) = row.into_typed::<(chrono::Duration, )>().next() {
            let row = row?.0;
            return Ok(Some(Generation(row)))
        }
    }

    Ok(None)
}
```